modules/core/src/main/java/org/apache/ignite/internal/jdbc2/JdbcBinaryBuffer.java

```diff
@@ -0,0 +1,463 @@
+     */
+    int read(int pos, byte[] resBuf, int resOff, int resLen) {
+        Objects.checkFromIndexSize(resOff, resLen, resBuf.length);
+
+        if (pos >= length)
+            return -1;
+
+        int bufOff = pos + off;
+
+        int size = Math.min(resLen, length - pos);
+
+        U.arrayCopy(arr, bufOff, resBuf, resOff, size);
+
+        return size;
+    }
+
+    /**
+     * Writes {@code inpLen} bytes from the specified byte array {@code bytes} starting at offset {@code off}
+     * to this storage to specified position {@code pos}.
+     *
+     * @param pos Pointer to a position.
+     * @param inpBuf Input byte array.
+     * @param inpOff Start offset in the input array.
+     * @param inpLen Number of bytes to write.
+     */
+    void write(int pos, byte[] inpBuf, int inpOff, int inpLen) throws IOException {
+        if (MAX_ARRAY_SIZE - pos < inpLen)
+            throw new IOException("Too much data. Can't write more then " + MAX_ARRAY_SIZE + " bytes.");
+
+        Objects.checkFromIndexSize(inpOff, inpLen, inpBuf.length);
+
+        int newLen = Math.max(pos + inpLen, length);
+
+        ensureCapacity(newLen);
+
+        U.arrayCopy(inpBuf, inpOff, arr, pos, inpLen);
+
+        length = newLen;
+    }
+
+    /**
+     * Read a byte from this buffer from specified position {@code pos}.
+     *
+     * @param pos Position.
+     * @return Byte read from the Blob. -1 if end of data reached.
+     */
+    int read(int pos) {
+        if (pos >= length)
+            return -1;
+
+        return arr[pos + off] & 0xff;
+    }
+
+    /**
+     * Write a byte to this buffer to specified position {@code pos}.
+     *
+     * <p>The byte to be written is the eight low-order bits of the
+     * argument {@code b}. The 24 high-order bits of {@code b}b are ignored.
+     *
+     * @param pos Pointer to a position.
+     * @param b Byte to write.
+     */
+    void write(int pos, int b) throws IOException {
+        if (MAX_ARRAY_SIZE - pos < 1)
+            throw new IOException("Too much data. Can't write more then " + MAX_ARRAY_SIZE + " bytes.");
+
+        int newLen = Math.max(pos + 1, length);
+
+        ensureCapacity(newLen);
+
+        arr[pos] = (byte)b;
+
+        length = newLen;
+    }
+
+    /**
+     * Ensure capacity.
+     *
+     * @param newLen The new data length the buffer should be able to hold.
+     */
+    private void ensureCapacity(int newLen) {
+        if (newLen - arr.length > 0 || isReadOnly)
+            reallocate(capacity(arr.length, newLen));
+    }
+
+    /**
+     * Calculate new capacity.
+     *
+     * @param curCap Current capacity.
+     * @param reqLen Required new data length.
+     * @return New capacity.
+     */
+    protected static int capacity(int curCap, int reqLen) {
+        int newCap;
+
+        if (reqLen < MIN_CAP)
+            newCap = MIN_CAP;
+        else {
+            newCap = Math.max(curCap, MIN_CAP);
+
+            while (newCap < reqLen) {
```

Comment:
Infinite loop on `reqLen == MAX_ARRAY_SIZE`?

Reply:
No.  

There is a unit test for that case

https://github.com/apache/ignite/pull/11518/files#diff-a36e9a0a75d46d9f20fb71c6f9de5d7d71f3e36004416838c00bfe2aa13d8c51R37